src/main/java/net/neoforged/neoforge/event/ValidityEvent.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.event;
+
+import java.util.Optional;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.entity.vehicle.ContainerEntity;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.SignBlockEntity;
+import net.neoforged.neoforge.common.NeoForge;
+import net.neoforged.neoforge.event.entity.player.PlayerEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Event to check if a menu can stay open or if the player has moved too far away from it.
+ * <p>
+ * This is intended for mods that allow a player remote access to blocks and their menus,
+ * but it also can be used to forcibly kick them out of menus.
+ * <p>
+ * You can query the result of the original distance check with {@link #getOriginalValue()}
+ * and set your override with {@link #setResult(boolean)}.
+ * <p>
+ * This base class is never posted (but it can be subscribed to, as it is not abstract),
+ * only one of its subclasses:
+ * <ul>
+ * <li>{@link ValidityEvent.Menu} for all kinds of menus,
+ * <li>{@link ValidityEvent.EntityMenu} for menus attached to an entity, or
+ * <li>{@link ValidityEvent.ContainerEntityMenu} for {@link ContainerEntity}-based menus, or
+ * <li>{@link ValidityEvent.Sign} for the checks done by signs which don't use menus but a Screen.
+ * </ul>
+ * This event is fired on the {@link NeoForge#EVENT_BUS}.
+ */
+public class ValidityEvent extends PlayerEvent {
```

Comment:
The event children is pretty complicated due to inlining the original checking method in the event constructor.
Since we only holds the original boolean result anyways, just pass in the boolean original, it's the subscriber's choice to decide how they would modify the boolean, and they should be aware of the original check if they are only subscribing for certain entities.